pkg/prometheus/agent/operator.go

```diff
@@ -954,15 +954,25 @@ func (c *Operator) createOrUpdateConfigurationSecret(ctx context.Context, p *mon
 	}
 
 	// Update secret based on the most recent configuration.
-	conf, err := cg.GenerateAgentConfiguration(
-		smons,
-		pmons,
-		bmons,
-		scrapeConfigs,
-		p.Spec.TSDB,
-		store,
-		additionalScrapeConfigs,
-	)
+	var conf []byte
+	if ptr.Deref(p.Spec.Mode, "StatefulSet") == "DaemonSet" {
+		conf, err = cg.GenerateAgentDaemonSetConfiguration(
```

Comment:
I'm not sure if we need separate functions `GenerateAgentDaemonSetConfiguration` and `GenerateAgentConfiguration` here. The object that calls them, `cg`, has `PrometheusInterface` as one of its attributes, which means that `cg` is able to tell if it is generating the configuration to a daemonset or not.


https://github.com/prometheus-operator/prometheus-operator/blob/bf3a738f8ce631035005884f99ed1f1a7ef453a2/pkg/prometheus/promcfg.go#L2558

Here we could do something like:

```go
	if cg.prom.(*monitoringv1alpha1.PrometheusAgent).Spec.Mode == "Daemonset" {
		scrapeConfigs = cg.appendPodMonitorDaemonsetConfigs(scrapeConfigs, pMons, apiserverConfig, store, shards)
	} else {
		scrapeConfigs = cg.appendPodMonitorConfigs(scrapeConfigs, pMons, apiserverConfig, store, shards)
	}
```

and avoid a lot of duplicated code we have in this PR

Reply:
Yeah a know the PR is a bit messy at this moment^^" as I said in https://github.com/prometheus-operator/prometheus-operator/pull/6879#issuecomment-2322821270 I'll clean it up before we merge

Can you help me with the error in https://github.com/prometheus-operator/prometheus-operator/pull/6879#discussion_r1741078817 too? This error is about not being able to retrieve our sample app's pods, which doesn't really relate to podmonitor logic anyway

cc @simonpasquier